### ISSUE TITLE
cached_properties as valid implementations of properties

### ISSUE THREAD

**Issue by tdsmith** [microsoft/pyright#4343]
**Is your feature request related to a problem? Please describe.**
Implementations that use functools.cached_property are regarded as incompatible with interfaces that declare a property.

**Describe the solution you'd like**
This would check without error:

```python
from functools import cached_property
from typing import Protocol

class A(Protocol):
    @property
    def foo(self) -> int:
        ...

class B:
    @cached_property
    def foo(self) -> int:
        return 42

a: A = B()
```

Presently, this fails at the last line, with the message

```
Expression of type "B" cannot be assigned to declared type "A"
  "B" is incompatible with protocol "A"
    "foo" is an incompatible type
      "cached_property[int]" is incompatible with "int"
```

I think I would like cached_properties to be treated as a kind of property.

**Additional context**

A workaround which pyright accepts is `cached_property = cast(Type[property], functools.cached_property)` -- though of course this doesn't cover cached_property-specific interfaces like deletion (`del B().foo`) to clear the cached response.

**Comment by erictraut** [comment]
Yeah, this is unfortunate. The `property` class necessarily requires a bunch of special-case logic in type checkers. By comparison, `cached_property` is modeled as a normal class and has special properties defined in typeshed.

I don't see a good way to unify the two, unfortunately. You'll need to continue to use hacks like the one you mentioned above.

**Comment by tdsmith** [comment]
Roger, thanks!
